01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+  track them all as outputs of a build. This could also support batch creation.
+
+  To ensure `check` is the source of truth for ordering, the versions returned
+  by `put` are not order dependent. A `check` will be performed to discover
+  them in the correct order, and then each version will be saved as an output
+  of the build. The latest version of the set will then be fetched.
+
+* Change `put` to write its JSON response to a specified file, rather than
+  `stdout`, so that we don't have to be attached to process its response.
+
+  This is one of the few ways a build can error after the ATC reattaches
+  (`unexpected end of JSON`). With it written to a file, we can just try to read
+  the file when we re-attach after seeing that the process exited. This also
+  frees up stdout/stderr for normal logging, which has been an occasional
+  pitfall during resource development/debugging.
+
+
+## Introduction of "notifier" resource interface
+
+A resource may now implement a "notifier" interface. Notifications must be
+explicitly enabled in the pipeline. This is so that a resource type can't
+suddenly decide to include a notifier and start doing surprising things.
+
+A notifier has a single hook, `notify`, and receives some sort of JSON payload
+describing the event. The event is purely for side-effects and has no
+implications for pipeline semantics.
+
+If a notification fails to send, the build shall error. To be honest, I haven't
+thought about this much yet, but I think it's better to be conservative.
+
+A resource type may implement both "artifacts" and "notifier" interfaces. If a
+resource implements both, it will be given information about its occurrence in
+the build (space and version). This will be crucial for e.g. reporting the
+status of a pull request or commit back to GitHub.
+
+Add support for a "notifier" resource type. They have one hook, `notify`,
+which is invoked with various types of notifications
+(`build_started`, `build_finished`, as the poster children).
+
+
+# Caveats
+
+Terminology is now even more confusing. "Resource type" could mean either the
+"git" vs. "s3" or "notifier" vs. "artifact".
+
+
+# Open Questions
+
+## Can we reduce the `check` overhead?
+
+With spaces there will be more `check`s than ever. Right now, there's one
+container per recurring `check`. Can we reduce the container overhead here by
+requiring that resource `check`s be side-effect free and able to run in parallel?
+
+There may be substantial security implications for this.
+
+
+## Is `destroy` general enough to be a part of the interface?
+
+It may be the case that most resources cannot easily support `destroy`. One
+example is the `git` resource. It doesn't really make sense to `destroy` a
+commit. Even if it did (`push -f`?), it's a kind of weird workflow to support
+out of the box.
+
+Could we instead just have `put` and ensure that we `check` in such a way that
+deleted versions are automatically noticed? What would the overhead of this be?
```

Comment:
the semantics for how to implement this in `check` sound difficult... does check now both get and return a list of a deleted resources? how am i supposed to figure out what those are, especially after they've been deleted?

Reply:
Yeah, this part's fairly tricky. Here is one way I've thought of:

When `check` is given a version that's no longer available, it could indicate that it's gone by returning the latest version *plus* the version prior to it. Because the cursor version isn't returned, it would be marked as 'deleted', and any existing versions in the database that were previously *in between* the returned versions would be marked as deleted as well.

This would work for `git` because there's an inherent linearity to its history wherein its commits point directly to a parent commit, but I haven't reasoned about it for other implementations or thought about other more complicated situations.

Reply:
even in the git case, what if I delete HEAD?

Reply:
In the git case, if you delete `HEAD` (assuming you mean `push -f` of the commit prior to it), the version that was *previously* at `HEAD` would be given to `check` (since all checks run from a cursor, except for the initial one).

Today's resource interface says that if the given version is no longer there, it shouldn't be returned - instead, the new "latest" version should be returned.

So, assuming you started from this history:

```
commit 1
commit 2
commit 3
commit 4 <- HEAD
```

If you were to `push -f` this new history:

```
commit 1
commit 2
commit 3.1
commit 4.1
```

...then today, the the `check` script would return:

```
commit 4.1
```

And we could thus know that at least `commit 4` is no longer with us.

That leaves `commit 3` still there, though, so we haven't achieved *full* detection yet. This is where I'm stuck. Just thinking out loud, maybe the trick is to be able to `check` for ancestry in order to close the gap? Then the ATC could walk backwards in the history until either all prior versions are wiped out or until the loop is closed (by finding commit 4.1 -> commit 3.1 -> commit 2, which already exists, so the job would be done).

Reply:
Your example is slightly more difficult and more illustrative than what i meant initially... 

I think I meant if I `push -f` over your 4 commit example:
```
commit 1
commit 2
commit 3 <- HEAD
```
then check would get `commit 4` and it'd return `commit 2; commit 3`?

Reply:
(re-posting 'cause email reply doesn't keep context)

To be honest I wrote that it could include the prior version earlier but I don't think it helps enough. I think the ancestral scan would be necessary instead.

So in this case it would just return `commit 3` and Concourse would know that the given version is deleted (since it wasn't returned).

Reply:
In the resources I've written, the first `check` usually has to fetch multiple "potential" versions, figure out which is the last, return that (last) version and discard the rest. Would it make sense that `check` returns the complete list of _all_  (or 100-1000) versions on the first check (i.e. when it's passed an empty version)? So if we notice that a version is missing like in the scenario above, it just resets and start over like it was the first check again?

Reason I'm bringing this up is that it sounds like the "ancestral scan" would involve a lot of API calls to iterate backwards if someone decided to rebase `concourse/concourse` with random squashes over 8000 commits.

If the goal is to keep a complete historical record of all versions that passed through Concourse, this will obviously not do. But the way I see it, someone force pushing over a commit and us removing that version, does not change the fact that one or more artifacts might have been built on that particular revision of the code - so the get/put/build logs would still reference a version which no longer exists, which means they are our "source of truth".

Reply:
@itsdalmo Slight clarification: we would never remove versions, we would mark them deleted, and probably have them shown separately from the "source of truth" version history. This is purely to preserve those get/put associations. These 'soft deleted' versions wouldn't be available as inputs, they'd just be archived.

I like the simplicity of having resources just say "ok screw it here's everything I know" if the given version is gone, and diffing the history to notice which versions are gone. We'd have to be careful to design an API that's hard for resource authors to mess up, though. Everyone screws up pagination the first time. ;)